File: nodes/lcm_plan_dispatcher.py
#! /usr/bin/env python
import argparse
import attr
import yaml

# The lcm types
import lcm
from robotlocomotion.robot_plan_t import robot_plan_t

# The ros types
import rospy
import actionlib
from robot_msgs.msg import JointTrajectoryGoal, JointTrajectoryAction
from trajectory_msgs.msg import JointTrajectoryPoint


# The argument
parser = argparse.ArgumentParser()
parser.add_argument('--config_path',
                    default='',
                    help='The path to the configuration file')


@attr.s
class KukaPlanDispatcherConfig(object):
    # The ros channel
    joint_trajectory_action = '/plan_runner/JointTrajectory'
    stop_plan_service = '/plan_runner/stop_plan'
    # The lcm channel
    lcm_command_channel = 'COMMITTED_ROBOT_PLAN'
    lcm_stop_channel = 'STOP'


class KukaPlanDispatcherLCM(object):

    def __init__(self, config=KukaPlanDispatcherConfig()):
        self._config = config
        self._joint_trajectory_action = None  # type: actionlib.SimpleActionClient

    def start(self):
        # Setup of the action
        self._joint_trajectory_action = actionlib.SimpleActionClient(
            self._config.joint_trajectory_action,
            JointTrajectoryAction)
        self._joint_trajectory_action.wait_for_server()

        # The lcm staff
        lcm_handle = lcm.LCM()
        lcm_handle.subscribe(self._config.lcm_command_channel, self.handle_joint_trajectory_plan)

        # Start processing
        try:
            rate = rospy.Rate(5)
            while True:
                lcm_handle.handle()
                rate.sleep()
        except RuntimeError:
            pass

    def handle_joint_trajectory_plan(self, channel, data):
        msg = robot_plan_t.decode(data)
        if len(msg.plan) < 2 or msg.num_states < 2:
            return

        # Fill the joint name goal msg
        goal_msg = JointTrajectoryGoal()
        plan_0 = msg.plan[0]
        for i in range(0, len(plan_0.joint_name)):
            name_i = plan_0.joint_name[i]
            goal_msg.trajectory.joint_names.append(name_i)

        # Fill the plan
        for i in range(0, len(msg.plan)):
            plan_i = msg.plan[i]
            point_i = JointTrajectoryPoint()
            point_i.time_from_start = rospy.Duration(nsecs=plan_i.utime * 1000)
            for j in range(0, len(plan_i.joint_position)):
                point_i.positions.append(plan_i.joint_position[j])
                point_i.velocities.append(0)
                point_i.accelerations.append(0)
                point_i.effort.append(0)

            # Append to trajectory
            goal_msg.trajectory.points.append(point_i)

        # Send the goal
        self._joint_trajectory_action.send_goal(goal_msg)


def main():
    # Parse the argument
    args, unknown = parser.parse_known_args()
    config_path = args.config_path  # type: str

    # Construct the the config
    dispatcher_config = KukaPlanDispatcherConfig()
    if len(config_path) > 0:
        with open(config_path) as config_file:
            datamap = yaml.load(config_file, Loader=yaml.CLoader)
            dispatcher_config.lcm_command_channel = datamap['lcm_plan_channel']
            dispatcher_config.lcm_stop_channel = datamap['lcm_stop_channel']
            config_file.close()

    # OK
    rospy.init_node('lcm_command_bridge')
    dispatcher = KukaPlanDispatcherLCM(dispatcher_config)
    dispatcher.start()


if __name__ == '__main__':
    main()

File: src/kuka_lcm/lcm_plan_dispatcher.cpp
//
// Created by wei on 9/9/19.
//

#include <std_srvs/Trigger.h>

#include "lcm_plan_dispatcher.h"
#include "trajectory_msgs/JointTrajectoryPoint.h"
#include "robot_msgs/JointTrajectoryActionGoal.h"


plan_runner::KukaPlanDispatcherLCM::KukaPlanDispatcherLCM(
        ros::NodeHandle &nh,
        std::string lcm_plan_channel, std::string lcm_stop_channel,
        const std::string& joint_trajectory_action, const std::string& stop_plan_service
) : node_handle_(nh),
    lcm_plan_channel_(std::move(lcm_plan_channel)), lcm_stop_channel_(std::move(lcm_stop_channel))
{
    joint_trajectory_client_ = std::make_shared<actionlib::SimpleActionClient<robot_msgs::JointTrajectoryAction>>(
            joint_trajectory_action, true);
    stop_plan_client_ = std::make_shared<ros::ServiceClient>(node_handle_.serviceClient<std_srvs::Trigger>(stop_plan_service));
}


void plan_runner::KukaPlanDispatcherLCM::Start() {
    // Check the server
    joint_trajectory_client_->waitForServer();

    // Invoke lcm
    lcm::LCM lcm;
    lcm.subscribe(lcm_plan_channel_, &KukaPlanDispatcherLCM::handleJointTrajectoryPlan, this);
    lcm.subscribe(lcm_stop_channel_, &KukaPlanDispatcherLCM::handleStopPlan, this);

    // Spin
    while (0 == lcm.handle()) {};
}


void plan_runner::KukaPlanDispatcherLCM::handleJointTrajectoryPlan(
    const lcm::ReceiveBuffer *,
    const std::string &,
    const robotlocomotion::robot_plan_t *tape
) {
    // Basic check
    const auto& lcm_plan = *tape;
    if(lcm_plan.num_states < 2 || lcm_plan.plan.size() < 2) return;

    // Fill the joint name
    auto goal_msg = robot_msgs::JointTrajectoryGoal();
    for(size_t i = 0; i < lcm_plan.plan[0].joint_name.size(); i++) {
        auto name_i = lcm_plan.plan[0].joint_name[i];
        goal_msg.trajectory.joint_names.emplace_back(std::move(name_i));
    }

    // Fill the plan
    for(size_t i = 0; i < lcm_plan.plan.size(); i++) {
        const auto& state_i = lcm_plan.plan[i];
        trajectory_msgs::JointTrajectoryPoint point;
        for(size_t j = 0; j < state_i.joint_position.size(); j++) {
            point.positions.push_back(state_i.joint_position[j]);

            // Other properties are zero
            point.velocities.push_back(0);
            point.accelerations.push_back(0);
            point.effort.push_back(0);
        }

        // Append this point
        point.time_from_start.fromSec(double(state_i.utime) / 1e6);
        goal_msg.trajectory.points.emplace_back(std::move(point));
    }

    // Send to robot
    std::lock_guard<std::mutex> guard(mutex_);
    joint_trajectory_client_->sendGoal(goal_msg);
}

void plan_runner::KukaPlanDispatcherLCM::handleStopPlan(
    const lcm::ReceiveBuffer *,
    const std::string &,
    const robotlocomotion::robot_plan_t *
) {
    std_srvs::Trigger request;
    std::lock_guard<std::mutex> guard(mutex_);
    stop_plan_client_->call(request);
}
